syntax = "proto3";

import "app/models/proto/auth.proto";
import "app/models/proto/shared.proto";
import "buf/validate/validate.proto";

service SettingsSecurityService {
  rpc ChangePassword(ChangePasswordRequest) returns (ChangePasswordResponse) {}
  rpc SetupTotp(SetupTotpRequest) returns (SetupTotpResponse) {}
  rpc DisableTotp(DisableTotpRequest) returns (DisableTotpResponse) {}
  rpc RegisterPasskey(RegisterPasskeyRequest) returns (RegisterPasskeyResponse) {}
  rpc RenamePasskey(RenamePasskeyRequest) returns (RenamePasskeyResponse) {}
  rpc RemovePasskey(RemovePasskeyRequest) returns (RemovePasskeyResponse) {}
  rpc GenerateRecoveryCodes(GenerateRecoveryCodesRequest) returns (GenerateRecoveryCodesResponse) {}
  rpc RevokeToken(RevokeTokenRequest) returns (RevokeTokenResponse) {}
}

message ChangePasswordRequest {
  TransmitUserPassword old_password = 1 [(buf.validate.field).required = true];
  TransmitUserPassword new_password = 2 [(buf.validate.field).required = true];
  bool revoke_other_sessions = 3;
}

message ChangePasswordResponse {
  StandardFeedbackDetail feedback = 1;
}

message SetupTotpRequest {
  string secret = 1 [(buf.validate.field).string.min_len = 1];
  uint32 digits = 2 [(buf.validate.field).uint32 = {
    in: [
      6,
      8
    ]
  }];
  string totp_code = 3 [(buf.validate.field).string.pattern = "^(?:[0-9]{6}|[0-9]{8})$"];
}

message SetupTotpResponse {}

message DisableTotpRequest {
  TransmitUserPassword password = 1 [(buf.validate.field).required = true];
}

message DisableTotpResponse {}

message RegisterPasskeyRequest {
  PasskeyRegistration registration = 1 [(buf.validate.field).required = true];
}

message RegisterPasskeyResponse {}

message RenamePasskeyRequest {
  bytes credential_id = 1;
  string name = 2 [(buf.validate.field).string.max_len = 50];
}

message RenamePasskeyResponse {
  string name = 1;
}

message RemovePasskeyRequest {
  TransmitUserPassword password = 1 [(buf.validate.field).required = true];
  bytes credential_id = 2;
}

message RemovePasskeyResponse {}

message GenerateRecoveryCodesRequest {
  TransmitUserPassword password = 1 [(buf.validate.field).required = true];
}

message GenerateRecoveryCodesResponse {
  repeated string codes = 1;
}

message RevokeTokenRequest {
  uint64 token_id = 1 [(buf.validate.field).uint64.gt = 0];
}

message RevokeTokenResponse {}
